osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Osu.Objects.Drawables;
+using osu.Framework.Logging;
+
+namespace osu.Game.Rulesets.Osu.Skinning.Legacy
+{
+    public class LegacyFollowCircle : CompositeDrawable
+    {
+        private readonly Drawable animationContent;
+        private DrawableSlider slider;
+
+        public LegacyFollowCircle(Drawable animationContent)
+        {
+            this.animationContent = animationContent;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(DrawableHitObject drawableObject)
+        {
+            slider = (DrawableSlider)drawableObject;
+
+            RelativeSizeAxes = Axes.Both;
+
+            InternalChild = animationContent;
+            animationContent.Anchor = Anchor.Centre;
+            animationContent.Origin = Anchor.Centre;
+
+            slider.Tracking.BindValueChanged(trackingChanged, true);
+            slider.ApplyCustomUpdateState += updateStateTransforms;
+        }
+
+        private void trackingChanged(ValueChangedEvent<bool> e)
+        {
+            if (slider.Judged)
+                return;
+
+            bool tracking = e.NewValue;
+
+            if (slider.Ball.InputTracksVisualSize)
+            {
+                if (tracking)
+                    this.ScaleTo(DrawableSliderBall.FOLLOW_AREA, 200, Easing.OutQuint);
+                else
+                    this.ScaleTo(DrawableSliderBall.FOLLOW_AREA * 2, 100, Easing.OutQuint).Then().ScaleTo(1f);
+            }
```

Comment:
This is weird. The logic is conditioned upon `InputTracksVisualSize`. But the `DrawableSliderBall`'s input receptor does not scale in the same manner. So am I to understand input will actually *not* track visual size on legacy skins?

I almost think that skinnable components handling `InputTracksVisualSize` is wrong and it's actually `DrawableSliderBall` that should be figuring that logic out.

Reply:
(Might be misunderstanding the question?) Yeah, legacy skins won't match the actual follow area. In stable, there's an animation were the follow circle goes crazy huge when you let go, but I don't think the follow area scales outwards like that too (correct me if I'm wrong). Default/legacy skin will have different animations, and I don't think it makes sense to give them differently animated follow receptors. One of the two skins is going to have slightly unrepresentative animations. It's weird, but I couldn't think of a better solution.

Reply:
if the goal is to make animations work 1:1 (as opposed to actually improving the gameplay readability as was done before this PR), the legacy skin shouldn't be checking this variable at all.

i'm still not sure and not going to review this PR until i am, though.

Reply:
Given the explanation above I'm actually starting to (personal opinion follows) slightly oppose bringing back the legacy follow circle animation at *all*. It just complicates things needlessly and brings back some of the weird obscure mechanics that stable has, that in a nice world have no right to exist but people also got used to.

As things are right now on `master`, there are two simple options: either the follow area is exactly as you see it, or it is a constant size (equal to the one at end time). This pull makes it so that is the case kinda, only on the default skin, and on legacy skins the actual follow circle never follows the animation, but is either smaller than it and *still*  changing size in a manner invisible to the player, or constant size. Try explaining that one to a beginner player, it just sounds insane.

Reply:
That's pretty fair, but at the very least the follow circle fade out animation that plays at the end of a slider (not the one when the player lets go early) should follow old legacy animations, since they come after the slider has been judged so it doesn't matter if they match the follow area, right?

Reply:
Sure, I could agree with that part of it.